src/runner_metrics.py

```diff
@@ -0,0 +1,215 @@
+        workflow: The workflow name.
+        workflow_run_id: The workflow run id.
+        repository: The repository name.
+        event: The github event.
+    """
+
+    timestamp: NonNegativeFloat
+    workflow: str
+    workflow_run_id: str
+    repository: str
+    event: str
+
+
+class PostJobMetrics(BaseModel):
+    """Metrics for the post-job phase of a runner.
+
+    Args:
+        timestamp: The UNIX time stamp of the time at which the event was originally issued.
+        status: The status of the job.
+    """
+
+    timestamp: NonNegativeFloat
+    status: str
+
+
+class RunnerMetrics(BaseModel):
+    """Metrics for a runner.
+
+    Args:
+        installed_timestamp: The UNIX time stamp of the time at which the runner was installed.
+        pre_job: The metrics for the pre-job phase.
+        post_job: The metrics for the post-job phase.
+    """
+
+    installed_timestamp: NonNegativeFloat
+    pre_job: PreJobMetrics
+    post_job: Optional[PostJobMetrics]
+
+
+def _inspect_file_sizes(fs: shared_fs.SharedFilesystem) -> tuple[Path, ...]:
+    """Inspect the file sizes of the shared filesystem.
+
+    Args:
+        fs: The shared filesystem for a specific runner.
+
+    Returns:
+        A tuple of files whose size is larger than the limit.
+    """
+    files: list[Path] = [
+        fs.path.joinpath(PRE_JOB_METRICS_FILE_NAME),
+        fs.path.joinpath(POST_JOB_METRICS_FILE_NAME),
+        fs.path.joinpath(RUNNER_INSTALLED_TS_FILE_NAME),
+    ]
+
+    return tuple(
+        filter(lambda file: file.exists() and file.stat().st_size > FILE_SIZE_BYTES_LIMIT, files)
+    )
+
+
+def _extract_metrics_from_fs(fs: shared_fs.SharedFilesystem) -> Optional[RunnerMetrics]:
+    """Extract metrics from a shared filesystem.
+
+    Args:
+        fs: The shared filesystem for a specific runner.
+
+    Returns:
+        The extracted metrics if at least the pre-job metrics are present.
+
+    Raises:
+        CorruptMetricDataError: Raised if one of the files is not valid or too large.
+    """
+    if too_large_files := _inspect_file_sizes(fs):
+        raise CorruptMetricDataError(
+            f"File size of {too_large_files} is too large. "
+            f"The limit is {FILE_SIZE_BYTES_LIMIT} bytes."
+        )
+
+    try:
+        installed_timestamp = fs.path.joinpath(RUNNER_INSTALLED_TS_FILE_NAME).read_text()
+    except FileNotFoundError:
+        logger.exception("installed_timestamp not found for runner %s", fs.runner_name)
+        return None
+
+    logger.debug("Runner %s installed at %s", fs.runner_name, installed_timestamp)
+    try:
+        pre_job_metrics = json.loads(fs.path.joinpath(PRE_JOB_METRICS_FILE_NAME).read_text())
+        logger.debug("Pre-job metrics for runner %s: %s", fs.runner_name, pre_job_metrics)
+    except FileNotFoundError:
+        logger.warning("%s not found for runner %s.", PRE_JOB_METRICS_FILE_NAME, fs.runner_name)
```

Comment:
Minor/can be addressed in a follow-up PR: logger.exception instead?

Out of curiosity: whats the expected action/monitoring in case of metrics not found?

Reply:
This is not actually an error. This can happen if a runner never picks up a job (e.g. because the number of runners has been reduced).I think it is only necessary to take action if there are a lot of warnings, as this may indicate a programming error.